### DISCUSSION TITLE
Automatically configuring and enabling plugins after installing with `ark plugin:install`

### DISCUSSION THREAD

**Discussion by alessiodf** [ArkEcosystem/core#4591] [Q&A] [answered]
It seems counterintuitive that you can install a plugin via `ark plugin:install` but then nothing new happens when you restart Core 3.1.0 since you have to manually edit `app.json` to add the newly installed plugin there.

I see references to `PluginDiscoverer` in the latest commit. Is this the solution for this problem? If so, can you please explain how it will work from an end user perspective and also from the perspective of a plugin developer? For instance, if I develop a plugin that needs to start with the `relay` process only and not the `forger` or vice versa, how can I make that happen without expecting users to manually configure `app.json`? One idea could be to read a manifest file in the root of the plugin's directory to determine which processes (e.g. `core`, `relay`, `forger`, `snapshot`) the plugin should launch with. Or is it something I would have to check in the plugin code itself? Obviously I know it could be manually set in `app.json` but this is about avoiding the need for manual configuration.

**Comment by sebastijankuzner** [comment, marked as answer]
`ark plugin:install` will install plugin under the selected network data path (for example: /ark-core/devnet /plugins/). 

In order to use installed plugin it has to be included in the app.json. We cannot automatically load the plugin based on its manifest file, because we cannot determine which processes will use it (e.g. `core`, `relay`). The order of the plugin is also important, because plugins are loaded sequentially as they are listed in app.json. If you use dependencies to other packages that often means that it has to be loaded just after the dependent plugin and we cannot simplify it to be loaded after the other packages. For example if you want to decorate logger which is the first plugin, you want you package to be loaded right after it before other packages are loaded. 

In order to use plugin in the app.jso it has to export ServiceProvider class. You can look into core-manager plugin for reference ([ServiceProvider](https://github.com/ArkEcosystem/core-manager/blob/master/src/service-provider.ts), [export](https://github.com/ArkEcosystem/core-manager/blob/812ce7dedeb5e768a7179520ae83bdbbf6344633/src/index.ts#L3)).  

To avoid manual configuration your plugin can also export CLI [commands](https://github.com/ArkEcosystem/core-manager/blob/812ce7dedeb5e768a7179520ae83bdbbf6344633/src/index.ts#L1). You can implement custom command like `plugin_name:config` that will position plugin in the right place inside app.json. For reference take a look into [manager:config ](https://github.com/ArkEcosystem/core-manager/blob/master/src/commands/manager-config.ts)command.  Any additional ENV variables can also be set trough command. 

To sum it up you can implement the `CLI plugin` that extends CLI commands or `Core plugin` that exports ServiceProvider and can be set in app.json. Both principles can be used in the single package. You can check if command are registered properly via the `ark help` command. New command should be listed after plugin installation. 

If you are using only one network configuration CLI plugins are loaded automatically. When you are using multiple network configurations (devnet, testnet, mainnet) then provide desired network via `--network` flag.  For example `ark help --network=testnet`.

References:
- [Creating CLI Commands](https://ark.dev/docs/core/development/cli/create)
- [Core Plugins](https://ark.dev/docs/core/development/plugins/intro)

**Reply by alessiodf** [reply]
Thanks @sebastijankuzner, I had no idea you could add commands to the CLI. That's great.
